### PR TITLE
don't require that an IdP supply a specific text string to indicate that the user is not authenticated

### PR DESCRIPTION
**Pull request by lloyd**
this is for issue #2339.  As part of the fix I wanted to ensure that errors not originating from the primary got shown under extended error information in the dialog.

Note to reviewer: I am not convinced my method of propagating error information is How We Do It.  Please review.

### CODE REVIEW

File: example/primary/provision.html [hunk 1]
```diff
@@ -19,7 +19,7 @@
     $.get('/api/whoami')
       .success(function(who) {
         if (user != who) {
-          return navigator.id.raiseProvisioningFailure('user is not authenticated as target user');
+          return navigator.id.raiseProvisioningFailure('not authed');
```
Comment (lloyd) [inline example/primary/provision.html:22]: a tweak to the built-in primary to change the not-authed string

File: resources/static/common/js/user.js [hunk 2]
```diff
@@ -529,14 +528,16 @@ BrowserID.User = (function() {
           complete(userInfo);
         },
         function(error) {
-          if (error.code === "primaryError" && error.msg === "user is not authenticated as target user") {
+          // issue #2339 - in case an error is raised we don't care
+          // about the specific error code.
+          if (error.code === "primaryError") {
```
Comment (lloyd) [inline resources/static/common/js/user.js:533]: `.code` of primaryError implies we actually received a response from the primary - that their javascript loads ok and calls the functions we provide.  As long as that is the case we assume that the user is not authenticated.

Recall the flow here:
1. dialog load
2. 1st provisioning attempt
3. if fails, direct to authentication screen
4. upon redirect back, 2nd provisioning attempt.  

So this change means if `.raiseProvisioningFailure` is called on the first attempt, we'll always go to authentication.  Before, this was only if the IdP provided the exact string `user is not authenticated as target user`

File: resources/static/dialog/js/modules/authenticate.js [hunk 3]
```diff
@@ -78,7 +78,13 @@ BrowserID.Modules.Authenticate = (function() {
     else {
       showHint("addressInfo");
       user.addressInfo(email, onAddressInfo,
-        self.getErrorDialog(errors.addressInfo));
+                       function(info) {
+                         var ai = errors.addressInfo;
+                         if (info.detailedError) {
+                           ai = $.extend(ai, { message: info.detailedError });
+                         }
+                         self.getErrorDialog(ai)();
```
Comment (seanmonstar) [inline resources/static/dialog/js/modules/authenticate.js]: this doesn't pass the error as `lowLevelInfo`, and so we may miss out on network info?

Comment (lloyd) [inline resources/static/dialog/js/modules/authenticate.js]: I think I just need to pass `info` along, as that would be what used to happen.  right?  `self.getErrorDialog(ai)(info)`?

**Comment by shane-tomlinson** [comment]
@lloyd - I have been unable to get an error message to appear no matter what I do.

in `example/primary/provision.html`, I added the following line on line 17:

```
    return navigator.id.raiseProvisioningFailure('ghetto failure');
```

I expected this to show an error screen after authenticating with the example IdP, but so far have been unable to make this occur.

File: resources/static/dialog/js/modules/authenticate.js (hunk 3 above)
Comment (shane-tomlinson) [inline resources/static/dialog/js/modules/authenticate.js]: I think you can revert this change and simplify displaying the message by changing detailedError=>message in https://github.com/mozilla/browserid/pull/3081/files#L1R540

**Comment by lloyd** [comment]
@shane-tomlinson so with this change _never_ show an error to the user when the IdP calls `.raiseProvisioningFailure()`

Why?  Because if the primary is having trouble provisioning, _they_ will explain the problem to the user while the auth page is showing.

In cases where the primary is more hard core broken (any case where `fail()` is called here with an error other than `primaryError` https://github.com/mozilla/browserid/blob/dev/resources/static/common/js/provisioning.js) this means that the primary is too broken to catch their own errors, and we _will_ display an error msg to the user.

make sense?

File: resources/static/dialog/js/modules/authenticate.js (hunk 3 above)
Comment (shane-tomlinson) [inline resources/static/dialog/js/modules/authenticate.js]: I am wrong, my suggestion does not work. I need to give this some though because this seems like a common enough scenario where it should be taken care of automatically.

Comment (lloyd) [inline resources/static/dialog/js/modules/authenticate.js]: Yeah, seems like we should have an easy way to surface specific errors to make user error reports actionable.

Should that block this pr?

-- lloyd (thumb-typing)

Comment (shane-tomlinson) [inline resources/static/dialog/js/modules/authenticate.js]: I have a minimal change (22 insertions, 20 deletions - with tests) that makes defining action.message much easier.

**Comment by shane-tomlinson** [comment]
One final check we should make - ensure that any text that comes from the IdP is escaped before they XSS users.

**Comment by shane-tomlinson** [comment]
r+ from my side, a review of the changes I made would be great.

**Comment by shane-tomlinson** [comment]
+1. I dig it.
